File: Disjoint Set/Python/Redundant Connection.py
# Problem link: https://leetcode.com/problems/redundant-connection/description/
# Time: O(n)
# Space: O(n)
class disjointSet:

    def __init__(self, num):
        self.rank = [0] * (num + 1)
        self.parent = [index for index in range(num + 1)]

    def find(self, node):
        if node == self.parent[node]:
            return node
        self.parent[node] = self.find(self.parent[node])
        return self.parent[node]
    
    def union(self, nodeOne, nodeTwo):
        parentOne, parentTwo = self.find(nodeOne), self.find(nodeTwo)

        if parentOne == parentTwo:
            return
        
        rankOne, rankTwo = self.rank[parentOne], self.rank[parentTwo]

        if rankOne > rankTwo:
            self.parent[parentTwo] = parentOne
        elif rankOne < rankTwo:
            self.parent[parentOne] = parentTwo
        else:
            self.rank[parentOne] += 1
            self.parent[parentTwo] = parentOne

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        DS = disjointSet(len(edges))
        result = 0

        for index, (u, v) in enumerate(edges):
            if DS.find(u) == DS.find(v):
                result = index
            DS.union(u, v)
        
        return edges[result]

File: Disjoint Set/Java/Redundant Connection.java
// Problem link: https://leetcode.com/problems/redundant-connection/description/
// Time: O(n)
// Space: O(n)
class disjointSet {

    int[] rank;
    int[] parent;

    public disjointSet(int num) {
        this.rank = new int[num + 1];
        this.parent = new int[num + 1];

        for(int index = 0; index <= num; index++) {
            this.parent[index] = index;
        }
    }

    public int find(int node) {
        if(node == this.parent[node]) {
            return node;
        }

        this.parent[node] = this.find(this.parent[node]);
        return this.parent[node];
    }

    public void union(int nodeOne, int nodeTwo) {
        int parentOne = this.parent[nodeOne], parentTwo = this.parent[nodeTwo];

        if(parentOne == parentTwo) {
            return;
        }

        int rankOne = this.rank[parentOne], rankTwo = this.rank[parentTwo];

        if(rankOne > rankTwo) {
            this.parent[parentTwo] = parentOne;
        }
        else if(rankOne < rankTwo) {
            this.parent[parentOne] = parentTwo;
        }
        else {
            this.rank[parentOne] += 1;
            this.parent[parentTwo] = parentOne;
        }
    }
}

class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int result = 0;
        disjointSet DS = new disjointSet(edges.length);

        for(int index = 0; index < edges.length; index++) {
            if(DS.find(edges[index][0]) == DS.find(edges[index][1])) {
                result = index;
            }
            DS.union(edges[index][0], edges[index][1]);
        }

        return edges[result];
    }
}
